managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/SharedEntryCacheManagerImpl.java

```diff
@@ -0,0 +1,218 @@
+    }
+
+    @Override
+    public EntryCache getEntryCache(ManagedLedgerImpl ml) {
+        if (getMaxSize() > 0) {
+            return new SharedEntryCacheImpl(ml, this);
+        } else {
+            return new EntryCacheDisabled(ml);
+        }
+    }
+
+    @Override
+    public void removeEntryCache(String name) {
+        // no-op
+    }
+
+    @Override
+    public long getSize() {
+        long totalSize = 0;
+        for (int i = 0; i < segmentsCount; i++) {
+            totalSize += segments.get(i).getSize();
+        }
+        return totalSize;
+    }
+
+    @Override
+    public long getMaxSize() {
+        return config.getMaxCacheSize();
+    }
+
+    @Override
+    public void clear() {
+        segments.forEach(SharedCacheSegment::clear);
+    }
+
+    @Override
+    public void close() {
+        segments.forEach(SharedCacheSegment::close);
+    }
+
+    @Override
+    public void updateCacheSizeAndThreshold(long maxSize) {
+
+    }
+
+    @Override
+    public void updateCacheEvictionWatermark(double cacheEvictionWatermark) {
+        // No-Op. We don't use the cache eviction watermark in this implementation
+    }
+
+    @Override
+    public double getCacheEvictionWatermark() {
+        return config.getCacheEvictionWatermark();
+    }
+
+    boolean insert(EntryImpl entry) {
+        int entrySize = entry.getLength();
+
+        if (entrySize > segmentSize) {
+            log.debug("entrySize {} > segmentSize {}, skip update read cache!", entrySize, segmentSize);
+            return false;
+        }
+
+        long stamp = lock.readLock();
+        try {
+            SharedCacheSegment s = segments.get(currentSegmentIdx);
+
+            if (s.insert(entry.getLedgerId(), entry.getEntryId(), entry.getDataBuffer())) {
+                return true;
+            }
+        } finally {
+            lock.unlockRead(stamp);
+        }
+
+        // We could not insert in segment, we to get the write lock and roll-over to
+        // next segment
+        stamp = lock.writeLock();
+
+        try {
+            SharedCacheSegment segment = segments.get(currentSegmentIdx);
+
+            if (segment.insert(entry.getLedgerId(), entry.getEntryId(), entry.getDataBuffer())) {
+                return true;
+            }
+
+            // Roll to next segment
+            currentSegmentIdx = (currentSegmentIdx + 1) % segmentsCount;
+            segment = segments.get(currentSegmentIdx);
+            segment.clear();
+            return segment.insert(entry.getLedgerId(), entry.getEntryId(), entry.getDataBuffer());
+        } finally {
+            lock.unlockWrite(stamp);
+        }
+    }
+
+    EntryImpl get(long ledgerId, long entryId) {
+        long stamp = lock.readLock();
+
+        try {
+            // We need to check all the segments, starting from the current one and looking
+            // backward to minimize the checks for recently inserted entries
+            for (int i = 0; i < segmentsCount; i++) {
```

Comment:
How about maintaining a mapping from the position <LedgerId, EntryId> to the SharedCacheSegment? 
With this mapping, we can get the target segment directly instead of traversing the whole segment list.